sources/org.osbuild.httpx

```diff
@@ -0,0 +1,183 @@
+a subscriptions.
+
+Internally use httpx to download the files; the files are cached in
+an internal cache. Multiple parallel connections are used to speed
+up the download.
+"""
+
+from typing import Dict
+
+import os
+import sys
+import tempfile
+import asyncio
+
+import httpx
+import aiofiles
+
+from osbuild import sources
+
+from osbuild.util.checksum import verify_file
+from osbuild.util.rhsm import Subscriptions
+
+
+SCHEMA = """
+"additionalProperties": false,
+"definitions": {
+  "item": {
+    "description": "The files to fetch indexed their content checksum",
+    "type": "object",
+    "additionalProperties": false,
+    "patternProperties": {
+      "(md5|sha1|sha256|sha384|sha512):[0-9a-f]{32,128}": {
+        "oneOf": [
+          {
+            "type": "string",
+            "description": "URL to download the file from."
+          },
+          {
+            "type": "object",
+            "additionalProperties": false,
+            "required": [
+              "url"
+            ],
+            "properties": {
+              "url": {
+                "type": "string",
+                "description": "URL to download the file from."
+              },
+              "secrets": {
+                "type": "object",
+                "additionalProperties": false,
+                "required": [
+                  "name"
+                ],
+                "properties": {
+                  "name": {
+                    "type": "string",
+                    "description": "Name of the secrets provider."
+                  }
+                }
+              }
+            }
+          }
+        ]
+      }
+    }
+  }
+},
+"properties": {
+  "items": {"$ref": "#/definitions/item"},
+  "urls": {"$ref": "#/definitions/item"}
+},
+"oneOf": [{
+  "required": ["items"]
+}, {
+  "required": ["urls"]
+}]
+"""
+
+
+class HttpxSource(sources.SourceService):
+
+    content_type = "org.osbuild.files"
+
+    # Max workers refers to max concurrent downloads in the HttpxSource and was
+    # chosed based on some arbitrary benchmarks
+    max_workers = 32
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.subscriptions = None
+
+    def transform(self, checksum, desc):
+        url = desc
+        if not isinstance(url, dict):
+            url = {"url": url}
+
+        # check if url needs rhsm secrets
+        if url.get("secrets", {}).get("name") == "org.osbuild.rhsm":
+            # rhsm secrets only need to be retrieved once and can then be reused
+            if self.subscriptions is None:
+                self.subscriptions = Subscriptions.from_host_system()
+            url["secrets"] = self.subscriptions.get_secrets(url.get("url"))
+        return checksum, url
+
+    def download(self, items: Dict) -> None:
+        items = [self.transform(*i) for i in items.items() if not self.exists(*i)]
+
+        if not items:
+            return
+
+        asyncio.run(self.fetch_many(*zip(*items)))
```

Comment:
Yea, I was thinking about the organisation but you end up with some annoying things regarding the abstract base class, fetch_one can't be both sync and async at the same time so it can't easily be something that can be run in either the loop *or* a threadpool.

I don't think having the super class dictate what runtime to use is a good plan though I do feel the current httpx implementation needs to follow the base class better. I'll do an iteration on it.